### PR TITLE
Add Phase-2 Level-1 trigger ML MET emulator

### PR DESCRIPTION
**Pull request by jmduarte**
#### PR description:

  - Add ML-based MET to Phase-2 Level-1 trigger emulator
  - Requires L1METML external integrated here: https://github.com/cms-sw/cmsdist/pull/8901
  - Presented at L1 Tau-Jets-MET meetings, e.g. https://indico.cern.ch/event/1081765/contributions/4549653
  - This model represents a v1 baseline with post-training quantization. We expect to update it with a more realistic firmware implementation.
  - Model and expected performance plots shown here: https://github.com/cms-hls4ml/L1METML/blob/main/README.md
  - Currently enabled for testing; but plan to switch it off to run standard MET by default
  - L1T Offline PR: https://github.com/cms-l1t-offline/cmssw/pull/1193
  - Model external repo: https://github.com/cms-hls4ml/L1METML

#### PR validation:

 - Run on 40 ([/RelValTTbar_14TeV/CMSSW_14_0_0_pre1-PU_133X_mcRun4_realistic_v1_2026D98PU200-v1/GEN-SIM-DIGI-RAW](https://cmsweb.cern.ch/das/request?input=dataset%3D%2FRelValTTbar_14TeV%2FCMSSW_14_0_0_pre1-PU_133X_mcRun4_realistic_v1_2026D98PU200-v1%2FGEN-SIM-DIGI-RAW&instance=prod/global)) events, and results are consistent with PUPPI MET (and ML algo)

### CODE REVIEW

**Comment by epalencia** [comment]
Please test with https://github.com/cms-sw/cmsdist/pull/8901

**Comment by aloeliger** [comment]
@jmduarte on a slightly picky note I see you've added the model to the cms-hls4ml organization here: https://github.com/cms-hls4ml/L1METML, could you please add that into the description of the PR?

Has this emulator been discussed outside the L1 Tau-Jet-MET group?

**Comment by aloeliger** [comment]
> Any progress on this PR? @jmduarte @aloeliger

@srimanob There were a few comments on the L1 branch here: https://github.com/cms-l1t-offline/cmssw/pull/1193. L1 has also recently upgraded it's prototype branch for phase 2 which seems to have caused some conflicts, so some rebasing of the original branch may be necessary.

**Comment by epalencia** [comment]
Please test

**Comment by aloeliger** [comment]
+l1
- L1 review is complete.
- L1 has merged this as a part of a phase 2 prototype branch,

**Comment by epalencia** [comment]
ping @cms-sw/upgrade-l2

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1MetPfProducer.cc [hunk 1]
```diff
@@ -48,34 +52,80 @@ class L1MetPfProducer : public edm::global::EDProducer<> {
...
+  int EncodePdgId(int pdgId) const;
+
   void CalcMetHLS(std::vector<float> pt, std::vector<float> phi, reco::Candidate::PolarLorentzVector& metVector) const;
+  void CalcMlMet(std::vector<float> pt,
+                 std::vector<float> eta,
+                 std::vector<float> phi,
+                 std::vector<float> puppiWeight,
+                 std::vector<int> pdgId,
+                 std::vector<int> charge,
+                 reco::Candidate::PolarLorentzVector& metVector) const;
```
Comment (subirsarkar) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1MetPfProducer.cc:78]: Is it really necessary to copy the vectors and pass to CalcMlMet (or CalcMetHLS). Can one not pass the const reference to the vectors?

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1MetPfProducer.cc [hunk 2]
```diff
@@ -48,34 +52,80 @@ class L1MetPfProducer : public edm::global::EDProducer<> {
   const int invTableBits_ = 10;
```
Comment (subirsarkar) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1MetPfProducer.cc:52]: Where-ever possible, please use **constexpr** in place of **const**.

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1MetPfProducer.cc [hunk 3]
```diff
@@ -84,8 +136,68 @@ void L1MetPfProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::Even
   iEvent.put(std::move(metCollection));
```
Comment (subirsarkar) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1MetPfProducer.cc:136]: I have another suggestion. Sorry that I missed it last time. 
The line
**std::unique_ptr<std::vector<l1t::EtSum>> metCollection(new std::vector<l1t::EtSum>(0));**
above may be changed to
**auto metCollection = std::make_unique<std::vector<l1t::EtSum>>(0);**
which is preferred (C++14 onwards).

**Comment by aloeliger** [comment]
@jmduarte Just a ping on the last comment and then we can sign and finalize this.

**Comment by jmduarte** [comment]
Sorry for the delay, comments have now been addressed.

**Comment by aloeliger** [comment]
please test

**Comment by subirsarkar** [comment]
+Upgrade

**Pull request merged**